### PR TITLE
Fix intel64 support of `F.batch_normalization`

### PR DESCRIPTION
**Pull request by toslunar**
There are two kinds of intel64 support:
- to use intel64 routine; and
- not to be broken with other intel64 enhancements.

https://github.com/chainer/chainer/pull/8175/files#r329861075 missed the latter point.  For example, backward of `h = bn(x); y = F.relu(h)` may give `isinstance(h.grad, intel64.mdarray)` even if `x.array` is not.

Nonetheless `astype` can be avoided.

By the way,
- `_BNMode.can_use_ideep` has been too strict from #6456.  `intel64.inputs_all_ready` checks `x.dtype` but it seems intended to use intel64 if `(x.dtype, gamma.dtype) == (float16, float32)`.  https://github.com/chainer/chainer/blob/4f4dda143fafccd914345ac28d750f5be1a99ef8/chainer/functions/normalization/batch_normalization.py#L634
  - I don't fix this issue in this PR.
- #8175 correctly fixed the condition for `(x.dtype, gamma.dtype) == (float32, float64)`, where intel64 should not be used.
  - #8175 also allowed `gamma.dtype` to be less precise than `x.dtype`, which does not seem to be often used to me.

### CODE REVIEW

**Comment by niboshi** [comment]
>For example, backward of `h = bn(x); y = F.relu(h)` may give `isinstance(h.grad, intel64.mdarray)` even if `x.array` is not.

I first implemented #8216 in a way that this case is correctly handled, but I figured out that was not needed in refatoring the fix in #8175, so I simplied the code as in #8216.
The idea was to make it possible to use different `BatchNormalizationImpl`s between forward and backward, by checking `gy` (just as done by `impl_selector` for forward).
I don't think simply reverting #8216 is not a good idea.
Can I re-implement that?

**Comment by niboshi** [comment]
Let's revert #8216 first and rebase this PR on it.
Refactoring can be considered separately.

**Comment by toslunar** [comment]
Bugfix is done by #8272.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
Jenkins failure seems unrelated to this PR. LGTM.
